### PR TITLE
[Dockerfile] build speed improvements

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR introduces an experimental Docker build workflow. It addresses the following issues with current Docker build process.
- Long cache export time. It takes about 600 seconds to export cache to the remote docker registry. Instead, the new workflow keeps the cache locally in persistent volumes that can be attached to different GHA runners.
- Re-building the workspace and dependencies for each source change. The new workflow uses cargo-chef to pre-build dependencies into a cacheable docker layer.
- Most `RUN` commands now have a cache mount, so cargo builds and package installs can leverage local cache, without needing to download or compile dependencies.

This is "experimental" and will not replace our current workflow, but those who wish can invoke this workflow by adding the `CICD:experimental-build` and/or `CICD:experimental-forge`. We will experiment with this for a week or so and make a decision to swap with the current workflow.

### Test Plan

The magic 12 minute build after cache.

image

### CODE REVIEW

**Review by geekflyer — APPROVED** [review]
amazing work! 
looking forward to try this out more in main.

File: docker/experimental/build-node.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+FEATURES=${FEATURES:-""}
+
+echo "Building aptos-node"
+echo "PROFILE: $PROFILE"
+echo "FEATURES: $FEATURES"
+
+# Build and overwrite the aptos-node binary with features if specified
+if [ -n "$FEATURES" ]; then
+    echo "Building aptos-node with features ${FEATURES}"
+    (cd aptos-node && cargo build --profile=$PROFILE --features=$FEATURES "$@")
```
[the remark this answers is not in the document]
Comment (rustielin) [reply docker/experimental/build-node.sh:16]: This has to do with feature unification across the whole workspace. By building with the features in a separate step, we can prevent re-building everything above, and just re-build the `aptos-node` binary with the required features.

File: .github/actions/docker-setup/action.yaml [hunk 2]
```diff
@@ -39,10 +39,12 @@ runs:
       run: docker context create builders
 
     - name: setup docker buildx
-      uses: docker/setup-buildx-action@15c905b16b06416d2086efa066dd8e3a35cc7f98 # pin v2.4.0
+      uses: aptos-labs/setup-buildx-action@7952e9cf0debaf1f3f3e5dc7d9c5ea6ececb127e # pin v2.4.0
```
Comment (rustielin) [inline .github/actions/docker-setup/action.yaml:42]: nice! Just saw https://github.com/aptos-labs/setup-buildx-action/commit/7952e9cf0debaf1f3f3e5dc7d9c5ea6ececb127e 

Does this mean the cache is cleared after each job without this fix?

File: .github/workflows/experimental-docker-build-test.yaml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    steps:
+      - name: collect metadata
+        run: |
+          echo "GIT_SHA: ${{ env.GIT_SHA }}"
+          echo "TARGET_CACHE_ID: ${{ env.TARGET_CACHE_ID }}"
+    outputs:
+      gitSha: ${{ env.GIT_SHA }}
+      targetCacheId: ${{ env.TARGET_CACHE_ID }}
+
+  rust-images:
```
Comment (rustielin) [inline .github/workflows/experimental-docker-build-test.yaml:101]: Will this work with all our other build variants as well? e.g. `indexer`, `failpoints`, `performance` ?

**Review by rustielin — APPROVED** [review]
Awesome work! Changes mostly LGTM -- thanks for making sure this experimental setup is isolated. Left some comments and questions above, but stamping to unblock

File: docker/experimental/build-node.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (geekflyer) [reply docker/experimental/build-node.sh:16]: I think my question is more like:
1. why do we need to go into that subdirectory? I don't think the working directory really matters that much - even if you run cargo build in a subdirectory the whole build is still considered to be part of the parent cargo workspace incl. the lockfile etc. Afaik the working directory only influences which package is built by default. To be concrete I believe `cd aptos-node && cargo build` is equivalent to `cargo build -p aptos-node` and I don't understand why we using the former for the feature build and the latter for the default build.
2. Why do we use --locked for the default build but not for the feature build?

Basically I think the only difference between the if and else condition should be the `--features` flag being present or not, the rest should be the same.

File: .github/actions/docker-setup/action.yaml (hunk 2 above)
Comment (ibalajiarun) [reply .github/actions/docker-setup/action.yaml:42]: The builder gets deleted after each run, and if the `keep-state` flag is not passed, the state gets deleted too along with builder. I changed it to keep the state, so when we try to recreate the builder with the _same name_ we can reuse the state.

File: .github/workflows/experimental-docker-build-test.yaml (hunk 3 above)
Comment (ibalajiarun) [reply .github/workflows/experimental-docker-build-test.yaml:101]: I mostly copy-pasted the scripts from our current build flow, so it should just work. Once this lands, we can experiment add those variants as well.
